functional_programming_tests/test/with_dependency_injection.cpp

```diff
@@ -0,0 +1,528 @@
+  request->map.layout.dim[2].stride = 1;
+
+  request->map.data = {0, 0, 0, 0, 0, 0, 0, 0,  //
+                       0, 0, 0, 1, 0, 0, 0, 0,  //
+                       0, 0, 0, 1, 0, 0, 0, 0,  //
+                       0, 0, 1, 1, 1, 0, 0, 0,  //
+                       0, 0, 1, 0, 1, 1, 0, 0,  //
+                       0, 0, 1, 0, 0, 0, 0, 0,  //
+                       0, 0, 0, 0, 0, 0, 0, 0,  //
+                       0, 0, 0, 0, 0, 0, 0, 0};
+  return request;
+}
+
+// Create mock middleware handle
+struct MockMiddlewareHandle : public PathGenerator::MiddlewareHandle {
+  MOCK_METHOD(void, register_set_map_service, (SetMapServiceCallback),
+              (override));
+  MOCK_METHOD(void, register_generate_path_service,
+              (GeneratePathServiceCallback), (override));
+  MOCK_METHOD(void, log_info, (std::string const&), (override));
+  MOCK_METHOD(void, log_error, (std::string const&), (override));
+};
+
+TEST(PathGenerator, Ctor) {
+  // GIVEN a mock middleware handle
+  auto mw = std::make_unique<MockMiddlewareHandle>();
+
+  // THEN it should register the services
+  EXPECT_CALL(*mw, register_set_map_service(testing::_)).Times(1);
+  EXPECT_CALL(*mw, register_generate_path_service(testing::_)).Times(1);
+
+  // WHEN the path generator is constructed
+  auto const path_generator = PathGenerator{std::move(mw)};
+}
+
+TEST(PathGenerator, SetMap) {
+  // GIVEN a path generator
+  auto mw = std::make_unique<MockMiddlewareHandle>();
+  // Capture the callback so it can be called later
+  PathGenerator::MiddlewareHandle::SetMapServiceCallback callback;
+  ON_CALL(*mw, register_set_map_service(testing::_))
+      .WillByDefault(testing::SaveArg<0>(&callback));
+
+  auto const path_generator = PathGenerator{std::move(mw)};
+
+  // WHEN the set map service is called
+  auto const request = make_costmap();
+  auto response = std::make_shared<example_srvs::srv::SetMap::Response>();
+  callback(request, response);
+
+  // THEN the path generator should successfully set the map
+  EXPECT_TRUE(response->success.data);
+}
+
+TEST(PathGenerator, NoCostmap) {
+  // GIVEN a path generator with a costmap
+  auto mw = std::make_unique<MockMiddlewareHandle>();
+  // Capture the path callback so it can be called later
+  PathGenerator::MiddlewareHandle::GeneratePathServiceCallback path_callback;
+  ON_CALL(*mw, register_generate_path_service(testing::_))
+      .WillByDefault(testing::SaveArg<0>(&path_callback));
+
+  auto const path_generator = PathGenerator{std::move(mw)};
+
+  // WHEN the generate path service is called without a costmap
+  auto path_request = std::make_shared<example_srvs::srv::GetPath::Request>();
+  path_request->start.data = {0, 0};
+  path_request->goal.data = {0, 0};
+
+  auto path_response = std::make_shared<example_srvs::srv::GetPath::Response>();
+  path_callback(path_request, path_response);
+
+  // THEN the path generator should fail
+  EXPECT_FALSE(path_response->success.data);
+}
```

Comment:
Same as above, but for the get path service, right?

Reply:
yup